### PR TITLE
Combine ManifestReaders into one parameterized class

### PR DESCRIPTION
**Pull request by rdblue**
This refactors `ManifestReader` and `DeleteManifestReader` into the same parameterized class, `ManifestReader<DataFile | DeleteFile>`. Using two separate classes didn't match the write side, where there is one `ManifestWriter` that is parameterized, and prevented code paths from supporting both data and delete files. For example, a class to merge manifests should be agnostic to the type of manifest, as long as either data or delete manifests are consistently merged.

This commit is also included in #1098 because it is required for those changes. I will rebase that commit this is merged first.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/ManifestFiles.java [hunk 1]
```diff
@@ -133,13 +134,29 @@ public static DeleteManifestReader readDeleteManifest(ManifestFile manifest, Fil
     throw new UnsupportedOperationException("Cannot write manifest for table version: " + formatVersion);
   }
 
+  static ManifestReader<?> open(ManifestFile manifest, FileIO io) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/ManifestFiles.java:137]: Note that these aren't public.

File: core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java [hunk 2]
```diff
@@ -235,7 +235,7 @@ private void performRewrite(List<ManifestFile> currentManifests) {
               keptManifests.add(manifest);
             } else {
               rewrittenManifests.add(manifest);
-              try (ManifestReader reader = ManifestFiles.read(manifest, ops.io(), ops.current().specsById())
+              try (ManifestReader<DataFile> reader = ManifestFiles.read(manifest, ops.io(), ops.current().specsById())
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java:238]: Does this mean we can only do rewrites if there are no `DeleteFile`'s?

Comment (danielcweeks) [reply core/src/main/java/org/apache/iceberg/BaseRewriteManifests.java:238]: Never mind, I see that the rewrite is only operating on the data manifests and not all manifests.

**Comment by danielcweeks** [comment]
+1  This looks good to me.  I would point out that it's confusing in a number of places where it's necessary to distinguish between `deleteManifests` and `deletedManifests`, but I'm not sure what we can do to be more clear about the distinction.

**Comment by rdblue** [comment]
I'll start renaming `deletedManifests` to `removedManifests` to be more clear. Thanks for pointing this out.

**Pull request merged**